File: back/api/models/animal.py
from django.db import models #type: ignore
from api.models.shelter import Shelter
from .choices.animal_choices import ANIMAL_STATUS_CHOICES
from .generate_ulid import generate_ulid

class Animal(models.Model):
    STATUS = ANIMAL_STATUS_CHOICES
    animal_id = models.CharField(
        max_length=26,
        default=generate_ulid,
        editable=False,
        unique=True)
    shelter = models.ForeignKey(
        Shelter,
        on_delete=models.PROTECT,   # PROTECT evita borrar shelters con animales asociados
        related_name="animals",
    )
    status = models.CharField(
        max_length=20,
        choices=ANIMAL_STATUS_CHOICES,
        default='rescued'
    )

    name = models.CharField(max_length=100)
    breed = models.CharField(max_length=100)
    chip = models.CharField(max_length=15)
    microchipped = models.BooleanField(default=False)
    birth_date = models.DateField(null=True, blank=True)
    intake_date = models.DateField()
    outcome_date = models.DateField(null=True, blank=True)
    species = models.CharField(max_length=50, choices=[('dog', 'Dog'), ('cat', 'Cat'), ('other', 'Other')])
    gender = models.CharField(max_length=10, choices=[('male', 'Male'), ('female', 'Female'), ('unknown', 'Unknown')])
    size = models.CharField(max_length=20, choices=[('small', 'Small'), ('medium', 'Medium'), ('large', 'Large')], blank=True)
    weight = models.DecimalField(max_digits=5, decimal_places=2, null=True, blank=True)

    sterilized = models.BooleanField(default=False)
    vaccinated = models.BooleanField(default=False)
    medical_notes = models.TextField(blank=True)
    allergies = models.TextField(blank=True)
    last_vet_visit = models.DateField(null=True, blank=True)

    temperament = models.CharField(max_length=200, blank=True)  # Ej: "Calm, friendly with kids"
    # good_with_dogs = models.BooleanField(default=False)
    # good_with_cats = models.BooleanField(default=False)
    # good_with_children = models.BooleanField(default=False)
    behavior_notes = models.TextField(blank=True)

    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

File: front/src/app/components/form-generator/forms_scaffolders/animal.ts
import { Animal, CreateAnimalDto } from '@app/core/services/animals/animals.model';
import { FormConfig, FormFieldConfig } from '../interfaces';

const animalFields: FormFieldConfig[] = [
  {
    name: 'name',
    label: 'Nombre',
    type: 'text',
    placeholder: 'Luna',
    suffixIcon: 'pets',
    required: true,
  },
  {
    name: 'species',
    label: 'Especie',
    type: 'select',
    required: true,
    options: [
      { value: 'dog', label: 'Perro' },
      { value: 'cat', label: 'Gato' },
      { value: 'other', label: 'Otro' },
    ],
  },
  {
    name: 'breed',
    label: 'Raza',
    type: 'text',
    placeholder: 'Mestizo',
    required: true,
  },
  {
    name: 'gender',
    label: 'Sexo',
    type: 'select',
    required: true,
    options: [
      { value: 'male', label: 'Macho' },
      { value: 'female', label: 'Hembra' },
      { value: 'unknown', label: 'Desconocido' },
    ],
  },
  {
    name: 'status',
    label: 'Estado',
    type: 'select',
    required: true,
    initialValue: 'rescued',
    options: [
      { value: 'rescued', label: 'Rescatado / Admitido' },
      { value: 'quarantine', label: 'Cuarentena' },
      { value: 'shelter', label: 'En refugio' },
      { value: 'foster', label: 'Acogida' },
      { value: 'treatment', label: 'Tratamiento veterinario' },
      { value: 'rehabilitation', label: 'Rehabilitación' },
      { value: 'reserved', label: 'Reservado' },
      { value: 'adopted', label: 'Adoptado' },
      { value: 'returned', label: 'Devuelto' },
      { value: 'transport', label: 'En traslado' },
      { value: 'lost', label: 'Perdido' },
      { value: 'deceased', label: 'Fallecido' },
      { value: 'euthanized', label: 'Eutanasiado' },
      { value: 'pre_intake', label: 'Pendiente de ingreso' },
      { value: 'transferred', label: 'Transferido' },
      { value: 'trial_adoption', label: 'Adopción de prueba' },
    ],
  },
  {
    name: 'chip',
    label: 'Chip',
    type: 'text',
    placeholder: '981000000000000',
    required: true,
  },
  {
    name: 'intake_date',
    label: 'Fecha de ingreso',
    type: 'date',
    required: true,
  },
  {
    name: 'birth_date',
    label: 'Fecha de nacimiento',
    type: 'date',
  },
  {
    name: 'size',
    label: 'Tamaño',
    type: 'select',
    options: [
      { value: 'small', label: 'Pequeño' },
      { value: 'medium', label: 'Mediano' },
      { value: 'large', label: 'Grande' },
    ],
  },
  {
    name: 'weight',
    label: 'Peso (kg)',
    type: 'number',
    placeholder: '12.5',
  },
  {
    name: 'last_vet_visit',
    label: 'Última visita veterinaria',
    type: 'date',
  },
  {
    name: 'temperament',
    label: 'Temperamento',
    type: 'text',
    placeholder: 'Sociable y tranquilo',
  },
  {
    name: 'medical_notes',
    label: 'Notas médicas',
    type: 'textarea',
    placeholder: 'Información veterinaria relevante',
  },
  {
    name: 'allergies',
    label: 'Alergias',
    type: 'textarea',
    placeholder: 'Sin alergias conocidas',
  },
  {
    name: 'behavior_notes',
    label: 'Notas de comportamiento',
    type: 'textarea',
    placeholder: 'Convivencia, miedos, rutina...',
  },
  {
    name: 'microchipped',
    label: 'Microchipado',
    type: 'checkbox',
  },
  {
    name: 'sterilized',
    label: 'Esterilizado',
    type: 'checkbox',
  },
  {
    name: 'vaccinated',
    label: 'Vacunado',
    type: 'checkbox',
  },
];

export const animalFormConfig: FormConfig = {
  fields: animalFields.map((field) => ({ ...field })),
  submitButtonText: 'Crear animal',
};

export function buildAnimalFormConfig(animal: Animal, submitButtonText = 'Guardar cambios'): FormConfig {
  const initialValues: Partial<Omit<CreateAnimalDto, 'shelter'>> = {
    status: animal.state,
    name: animal.name,
    breed: animal.breed,
    chip: animal.chip,
    birth_date: animal.birthDate,
    intake_date: animal.intakeDate,
    outcome_date: animal.outcomeDate,
    species: animal.species,
    gender: animal.gender,
    size: animal.size,
    weight: animal.weightValue,
    microchipped: animal.microchipped,
    sterilized: animal.sterilized,
    vaccinated: animal.vaccinated,
    medical_notes: animal.medicalNotes,
    allergies: animal.allergies,
    last_vet_visit: animal.lastVetVisit,
    temperament: animal.temperament,
    behavior_notes: animal.behaviorNotes,
  };

  return {
    fields: animalFields.map((field) => ({
      ...field,
      initialValue: initialValues[field.name as keyof typeof initialValues] ?? field.initialValue,
    })),
    submitButtonText,
  };
}
